File: R4GFX/Source/color_icc.c
/* R4OS userland adapter. LittleCMS is internal; no raw CMS context escapes
 * this owner. Every context and transform owns a bounded caller arena. */
#include "lcms2_plugin.h"
#include <stdint.h>

extern void *r4gfx_icc_alloc(void *arena, uint32_t bytes);
extern void *r4gfx_icc_realloc(void *arena, void *old, uint32_t bytes);
extern void r4gfx_icc_error(void *arena, uint32_t error);

static void *allocate(cmsContext ctx, cmsUInt32Number n) { return r4gfx_icc_alloc(cmsGetContextUserData(ctx), n); }
static void release(cmsContext ctx, void *p) { (void)ctx; (void)p; }
static void *resize(cmsContext ctx, void *p, cmsUInt32Number n) { return r4gfx_icc_realloc(cmsGetContextUserData(ctx), p, n); }
static void *zero(cmsContext ctx, cmsUInt32Number n) {
    void *p = allocate(ctx, n); if (p) memset(p, 0, n); return p;
}
static void *array(cmsContext ctx, cmsUInt32Number count, cmsUInt32Number size) {
    if (size && count > UINT32_MAX / size) return NULL;
    return zero(ctx, count * size);
}
static void *duplicate(cmsContext ctx, const void *data, cmsUInt32Number n) {
    void *p = allocate(ctx, n); if (p && n) memcpy(p, data, n); return p;
}
static const cmsPluginMemHandler memory_plugin = {
    { cmsPluginMagicNumber, LCMS_VERSION, cmsPluginMemHandlerSig, NULL },
    allocate, release, resize, zero, array, duplicate
};
static void report(cmsContext ctx, cmsUInt32Number code, const char *text) {
    (void)text; r4gfx_icc_error(cmsGetContextUserData(ctx), code);
}
static cmsContext context(void *arena) {
    cmsContext ctx = cmsCreateContext((void *)&memory_plugin, arena);
    if (ctx) cmsSetLogErrorHandlerTHR(ctx, report);
    return ctx;
}
typedef struct {
    cmsContext context;
    cmsHPROFILE profile, working;
    cmsHTRANSFORM transform;
    const cmsToneCurve *calibration[3];
    unsigned gray_input;
} R4GfxIcc;

void r4gfx_icc_close(void *handle) {
    R4GfxIcc *state = handle;
    if (!state || !state->context) return;
    if (state->transform) cmsDeleteTransform(state->transform);
    if (state->working) cmsCloseProfile(state->working);
    if (state->profile) cmsCloseProfile(state->profile);
    cmsDeleteContext(state->context);
    memset(state, 0, sizeof(*state));
}

/* direction0: encoded RGB -> relative PCS XYZ(D50). direction1: reverse.
 * input bytes remain in the caller arena through close; no host I/O exists. */
int r4gfx_icc_open(void *arena, const void *data, uint32_t length, uint32_t direction,
                   uint32_t intent, uint32_t black_compensation, uint32_t calibration, void **out) {
    cmsContext ctx;
    R4GfxIcc *state;
    cmsProfileClassSignature kind;
    cmsUInt32Number flags = cmsFLAGS_NOCACHE | cmsFLAGS_NOOPTIMIZE;
    if (direction > 1 || intent > 3 || black_compensation > 1 || calibration > 1 || (calibration && !direction)) return -1;
    ctx = context(arena);
    if (!ctx) return -2;
    state = zero(ctx, sizeof(*state));
    if (!state) { cmsDeleteContext(ctx); return -2; }
    state->context = ctx;
    state->profile = cmsOpenProfileFromMemTHR(ctx, data, length);
    if (!state->profile) goto fail;
    kind = cmsGetDeviceClass(state->profile);
    state->gray_input = !direction && cmsGetColorSpace(state->profile) == cmsSigGrayData;
    if ((!state->gray_input && cmsGetColorSpace(state->profile) != cmsSigRgbData) ||
        (direction && kind != cmsSigDisplayClass) ||
        (!direction && kind != cmsSigDisplayClass && kind != cmsSigInputClass && kind != cmsSigColorSpaceClass)) goto fail;
    /* Transform creation owns ICC.1:2022 section 8.10.2 tag precedence,
     * including A2B0/B2A0 when the requested intent has no dedicated LUT.
     * cmsIsIntentSupported only tests dedicated CLUT or matrix/shaper tags
     * and would reject valid LUT-only profiles before this fallback. */
    state->working = cmsCreateXYZProfileTHR(ctx);
    if (!state->working) goto fail;
    if (black_compensation) flags |= cmsFLAGS_BLACKPOINTCOMPENSATION;
    state->transform = direction ?
        cmsCreateTransformTHR(ctx, state->working, TYPE_XYZ_FLT, state->profile, TYPE_RGB_FLT, intent, flags) :
        cmsCreateTransformTHR(ctx, state->profile, state->gray_input ? TYPE_GRAY_FLT : TYPE_RGB_FLT, state->working, TYPE_XYZ_FLT, intent, flags);
    if (!state->transform) goto fail;
    if (calibration && cmsIsTag(state->profile, cmsSigVcgtTag)) {
        cmsToneCurve **curves = cmsReadTag(state->profile, cmsSigVcgtTag);
        unsigned i;
        if (!curves) goto fail;
        for (i = 0; i < 3; ++i) {
            if (!curves[i] || !cmsIsToneCurveMonotonic(curves[i]) || cmsIsToneCurveDescending(curves[i])) goto fail;
            state->calibration[i] = curves[i];
        }
    }
    *out = state;
    return 0;
fail:
    r4gfx_icc_close(state);
    return -3;
}

int r4gfx_icc_apply(void *handle, const float *input, float *output, uint32_t pixels) {
    R4GfxIcc *state = handle;
    uint32_t pixel;
    unsigned channel;
    if (state->gray_input) {
        float gray[64];
        /* RGBA image decoders expand gray into equal RGB channels. Reject
         * contradictory colored data before touching the destination. */
        for (pixel = 0; pixel < pixels; ++pixel)
            if (input[pixel * 3] != input[pixel * 3 + 1] || input[pixel * 3] != input[pixel * 3 + 2]) return -1;
        for (pixel = 0; pixel < pixels;) {
            unsigned n = pixels - pixel, i;
            if (n > 64) n = 64;
            for (i = 0; i < n; ++i) gray[i] = input[(pixel + i) * 3];
            cmsDoTransform(state->transform, gray, output + pixel * 3, n);
            pixel += n;
        }
    } else cmsDoTransform(state->transform, input, output, pixels);
    if (state->calibration[0]) for (pixel = 0; pixel < pixels; ++pixel)
        for (channel = 0; channel < 3; ++channel)
            output[pixel * 3 + channel] = cmsEvalToneCurveFloat(state->calibration[channel], output[pixel * 3 + channel]);
    return 0;
}

/* Original R4OS profile construction using the public LCMS API. Values are
 * white xy, RGB xy and three decoding exponents. Gray retains its one-channel
 * ICC characterization even though the image container uses RGBA storage. */
int r4gfx_icc_generate(void *arena, uint32_t gray, uint32_t srgb_curve,
                      const double *values, void *output, uint32_t *length) {
    cmsContext ctx = context(arena);
    cmsHPROFILE profile = NULL;
    cmsToneCurve *curves[3] = { NULL, NULL, NULL };
    const cmsCIExyY white = { values[0], values[1], 1 };
    const cmsCIExyYTRIPLE primaries = {
        { values[2], values[3], 1 }, { values[4], values[5], 1 }, { values[6], values[7], 1 }
    };
    const double srgb[5] = { 2.4, 1.0 / 1.055, 0.055 / 1.055, 1.0 / 12.92, 0.04045 };
    cmsUInt32Number needed = 0;
    int result = -1;
    unsigned i, channels = gray ? 1 : 3;
    if (!ctx) return -2;
    for (i = 0; i < channels; ++i) {
        curves[i] = srgb_curve ? cmsBuildParametricToneCurve(ctx, 4, srgb) : cmsBuildGamma(ctx, values[8 + i]);
        if (!curves[i]) goto done;
    }
    profile = gray ? cmsCreateGrayProfileTHR(ctx, &white, curves[0]) : cmsCreateRGBProfileTHR(ctx, &white, &primaries, curves);
    if (profile && cmsSaveProfileToMem(profile, NULL, &needed) && needed <= *length && cmsSaveProfileToMem(profile, output, &needed)) {
        *length = needed;
        result = 0;
    }
done:
    if (profile) cmsCloseProfile(profile);
    for (i = 0; i < channels; ++i) if (curves[i]) cmsFreeToneCurve(curves[i]);
    cmsDeleteContext(ctx);
    return result;
}

/* Deterministic standard profiles, serialized entirely in caller memory.
 * kind0: sRGB, kind1: linear BT.709, kind2: linear BT.2020. */
int r4gfx_icc_builtin(void *arena, uint32_t kind, void *output, uint32_t *length) {
    cmsContext ctx = context(arena);
    cmsHPROFILE profile = NULL;
    cmsToneCurve *curve = NULL;
    cmsUInt32Number needed = 0;
    int result = -1;
    if (!ctx) return -2;
    if (kind == 0) profile = cmsCreate_sRGBProfileTHR(ctx);
    else if (kind == 1 || kind == 2) {
        const cmsCIExyY white = { 0.3127, 0.3290, 1.0 };
        const cmsCIExyYTRIPLE srgb = { { 0.64, 0.33, 1 }, { 0.30, 0.60, 1 }, { 0.15, 0.06, 1 } };
        const cmsCIExyYTRIPLE bt2020 = { { 0.708, 0.292, 1 }, { 0.170, 0.797, 1 }, { 0.131, 0.046, 1 } };
        cmsToneCurve *curves[3];
        curve = cmsBuildGamma(ctx, 1.0);
        curves[0] = curves[1] = curves[2] = curve;
        if (curve) profile = cmsCreateRGBProfileTHR(ctx, &white, kind == 1 ? &srgb : &bt2020, curves);
    }
    if (profile && cmsSaveProfileToMem(profile, NULL, &needed) && needed <= *length &&
        cmsSaveProfileToMem(profile, output, &needed)) {
        *length = needed;
        result = 0;
    }
    if (profile) cmsCloseProfile(profile);
    if (curve) cmsFreeToneCurve(curve);
    cmsDeleteContext(ctx);
    return result;
}

File: R4GFX/Source/color_icc.zig
//! RGB ICC v2/v4 transforms, including LUT profiles, via the pinned LittleCMS
//! core. A profile's explicit intent is checked; no matrix-only substitution.
const std = @import("std");
comptime {
    _ = @import("color_c.zig");
}
pub const max_profile_bytes = 4 * 1024 * 1024;
pub const max_arena_bytes = 64 * 1024 * 1024;
pub const max_profile_tags = 100; // Exact capacity of the pinned ICC engine.
pub const Error = error{ Invalid, Unsupported, Memory, Alias, NonFinite };
pub const Direction = enum(u32) { input = 0, output = 1 };
pub const Intent = enum(u32) { perceptual = 0, relative = 1, saturation = 2, absolute = 3 };
pub const Arena = struct {
    bytes: []align(16) u8,
    used: usize = 0,
    failed: bool = false,
    error_code: u32 = 0,
    generation: u64 = 0,
    active: ?*anyopaque = null,
    pub fn init(bytes: []align(16) u8) Error!Arena {
        if (bytes.len < 1024 or bytes.len > max_arena_bytes) return error.Invalid;
        return .{ .bytes = bytes };
    }
    fn allocate(self: *Arena, count: usize) ?*anyopaque {
        if (count == 0) return null;
        const size = std.mem.alignForward(usize, count, 16);
        if (size > self.bytes.len - self.used or 16 > self.bytes.len - self.used - size) {
            self.failed = true;
            return null;
        }
        const header = self.bytes[self.used..][0..16];
        std.mem.writeInt(u64, header[0..8], count, .little);
        std.mem.writeInt(u64, header[8..16], @intFromPtr(self), .little);
        const data = self.bytes.ptr + self.used + 16;
        self.used += size + 16;
        return data;
    }
};
fn arena(context: ?*anyopaque) ?*Arena {
    const ptr = context orelse return null;
    return @ptrCast(@alignCast(ptr));
}
export fn r4gfx_icc_alloc(context: ?*anyopaque, count: u32) callconv(.c) ?*anyopaque {
    const owner = arena(context) orelse return null;
    return owner.allocate(count);
}
export fn r4gfx_icc_realloc(context: ?*anyopaque, previous: ?*anyopaque, count: u32) callconv(.c) ?*anyopaque {
    const owner = arena(context) orelse return null;
    const old = previous orelse return owner.allocate(count);
    const start = @intFromPtr(owner.bytes.ptr);
    const address = @intFromPtr(old);
    if (address < start + 16 or address >= start + owner.used) return null;
    const header: *const [16]u8 = @ptrFromInt(address - 16);
    if (std.mem.readInt(u64, header[8..16], .little) != @intFromPtr(owner)) return null;
    const old_size = std.mem.readInt(u64, header[0..8], .little);
    if (old_size > start + owner.used - address) return null;
    const next = owner.allocate(count) orelse return null;
    const to: [*]u8 = @ptrCast(next);
    const from: [*]const u8 = @ptrCast(old);
    @memcpy(to[0..@min(old_size, count)], from[0..@min(old_size, count)]);
    return next;
}
export fn r4gfx_icc_error(context: ?*anyopaque, code: u32) callconv(.c) void {
    const owner = arena(context) orelse return;
    if (owner.error_code == 0) owner.error_code = code;
}
extern fn r4gfx_icc_open(*Arena, [*]const u8, u32, u32, u32, u32, u32, *?*anyopaque) callconv(.c) c_int;
extern fn r4gfx_icc_close(?*anyopaque) callconv(.c) void;
extern fn r4gfx_icc_apply(*anyopaque, [*]const f32, [*]f32, u32) callconv(.c) c_int;
extern fn r4gfx_icc_builtin(*Arena, u32, ?[*]u8, *u32) callconv(.c) c_int;
extern fn r4gfx_icc_generate(*Arena, u32, u32, *const [11]f64, [*]u8, *u32) callconv(.c) c_int;
fn overlaps(left: usize, n: usize, right: usize, m: usize) bool {
    return n != 0 and m != 0 and left < right +| m and right < left +| n;
}
const Tag = struct { signature: u32, offset: usize, size: usize };
fn tag(bytes: []const u8, index: usize) Tag {
    const record = bytes[132 + index * 12 ..][0..12];
    return .{ .signature = std.mem.readInt(u32, record[0..4], .big), .offset = std.mem.readInt(u32, record[4..8], .big), .size = std.mem.readInt(u32, record[8..12], .big) };
}
fn validateProfile(bytes: []const u8) Error!void {
    if (bytes.len < 132 or bytes.len > max_profile_bytes or std.mem.readInt(u32, bytes[0..4], .big) != bytes.len or
        !std.mem.eql(u8, bytes[36..40], "acsp") or (bytes[8] != 2 and bytes[8] != 4)) return error.Invalid;
    const count = std.mem.readInt(u32, bytes[128..132], .big);
    if (count == 0) return error.Invalid;
    if (count > max_profile_tags) return error.Unsupported;
    const table_end = 132 + @as(usize, count) * 12;
    if (table_end > bytes.len) return error.Invalid;
    for (0..count) |i| {
        const item = tag(bytes, i);
        if (item.offset < table_end or item.offset % 4 != 0 or item.offset > bytes.len or
            item.size < 8 or item.size > bytes.len - item.offset) return error.Invalid;
        for (0..i) |j| {
            const prior = tag(bytes, j);
            if (item.signature == prior.signature) return error.Invalid;
            if (item.offset == prior.offset and item.size == prior.size) continue;
            if (overlaps(item.offset, item.size, prior.offset, prior.size)) return error.Invalid;
        }
    }
    // ICC4.4 requires contiguous unique tag elements, with zero padding.
    // Older versions retain their historical layout allowance.
    if (bytes[8] == 4 and bytes[9] >= 0x40) {
        var next = table_end;
        while (next < bytes.len) {
            var found: ?Tag = null;
            for (0..count) |i| {
                const item = tag(bytes, i);
                if (item.offset == next) {
                    found = item;
                    break;
                }
            }
            const item = found orelse return error.Invalid;
            const end = item.offset + item.size;
            next = std.mem.alignForward(usize, end, 4);
            if (next > bytes.len or !std.mem.allEqual(u8, bytes[end..next], 0)) return error.Invalid;
        }
    }
}

pub const Profile = struct {
    handle: ?*anyopaque = null,
    memory: *Arena,
    generation: u64 = 0,
    pub fn open(memory: *Arena, bytes: []const u8, direction: Direction, intent: Intent, black_compensation: bool, calibration: bool) Error!Profile {
        if (bytes.len < 132 or bytes.len > max_profile_bytes or memory.used != 0 or memory.active != null or (calibration and direction != .output)) return error.Invalid;
        if (overlaps(@intFromPtr(bytes.ptr), bytes.len, @intFromPtr(memory.bytes.ptr), memory.bytes.len) or
            overlaps(@intFromPtr(bytes.ptr), bytes.len, @intFromPtr(memory), @sizeOf(Arena)) or
            overlaps(@intFromPtr(memory), @sizeOf(Arena), @intFromPtr(memory.bytes.ptr), memory.bytes.len)) return error.Alias;
        // The ICC engine deliberately skips some invalid optional entries.
        // Output policy must not silently accept a structurally broken profile.
        try validateProfile(bytes);
        memory.generation = std.math.add(u64, memory.generation, 1) catch return error.Invalid;
        memory.failed = false;
        memory.error_code = 0;
        // The source is copied before LCMS retains its memory I/O handler.
        const copy: [*]u8 = @ptrCast(memory.allocate(bytes.len) orelse return error.Memory);
        @memcpy(copy[0..bytes.len], bytes);
        var result: Profile = .{ .memory = memory, .generation = memory.generation };
        const rc = r4gfx_icc_open(memory, copy, @intCast(bytes.len), @intFromEnum(direction), @intFromEnum(intent), @intFromBool(black_compensation), @intFromBool(calibration), &result.handle);
        if (rc != 0) {
            memory.used = 0;
            return if (memory.failed) error.Memory else error.Unsupported;
        }
        memory.active = result.handle;
        return result;
    }
    pub fn close(self: *Profile) void {
        const handle = self.handle orelse return;
        if (self.generation != self.memory.generation or self.memory.active != handle) {
            self.handle = null;
            return;
        }
        r4gfx_icc_close(handle);
        self.handle = null;
        self.memory.active = null;
        self.memory.used = 0;
    }
    // Float triplets are encoded RGB on the device side and relative XYZ(D50)
    // on the PCS side. The caller owns alpha, luminance scaling and composition.
    pub fn apply(self: *const Profile, input: []const [3]f32, output: [][3]f32) Error!void {
        const handle = self.handle orelse return error.Invalid;
        if (self.generation != self.memory.generation or self.memory.active != handle) return error.Invalid;
        if (input.len != output.len or input.len > 16 * 1024 * 1024) return error.Invalid;
        const in_bytes = std.mem.sliceAsBytes(input);
        const out_bytes = std.mem.sliceAsBytes(output);
        if (overlaps(@intFromPtr(out_bytes.ptr), out_bytes.len, @intFromPtr(in_bytes.ptr), in_bytes.len) or
            overlaps(@intFromPtr(out_bytes.ptr), out_bytes.len, @intFromPtr(self.memory.bytes.ptr), self.memory.bytes.len) or
            overlaps(@intFromPtr(out_bytes.ptr), out_bytes.len, @intFromPtr(self), @sizeOf(Profile)) or
            overlaps(@intFromPtr(out_bytes.ptr), out_bytes.len, @intFromPtr(self.memory), @sizeOf(Arena))) return error.Alias;
        for (input) |rgb| for (rgb) |v| if (!std.math.isFinite(v)) return error.NonFinite;
        if (r4gfx_icc_apply(handle, @ptrCast(input.ptr), @ptrCast(output.ptr), @intCast(input.len)) != 0) return error.Invalid;
        // Complex profile elements can overflow even for finite RGB. Never
        // publish a successful transform containing NaN/Inf to quantizers.
        for (output) |rgb| for (rgb) |v| if (!std.math.isFinite(v)) return error.NonFinite;
    }
};
pub const Definition = struct {
    chromaticities: @import("color.zig").Chromaticities,
    gamma: [3]f64 = @splat(1), // Device -> linear decoding exponents.
    gray: bool = false,
    srgb_curve: bool = false,
};
pub fn generate(memory: *Arena, definition: Definition, output: []u8) Error!usize {
    if (memory.used != 0 or memory.active != null or output.len < 132 or output.len > max_profile_bytes) return error.Invalid;
    if (overlaps(@intFromPtr(output.ptr), output.len, @intFromPtr(memory.bytes.ptr), memory.bytes.len) or
        overlaps(@intFromPtr(output.ptr), output.len, @intFromPtr(memory), @sizeOf(Arena)) or
        overlaps(@intFromPtr(memory), @sizeOf(Arena), @intFromPtr(memory.bytes.ptr), memory.bytes.len)) return error.Alias;
    const chroma = definition.chromaticities;
    if (!definition.gray) {
        _ = chroma.xyz() catch return error.Invalid;
    } else if (!std.math.isFinite(chroma.white[0]) or !std.math.isFinite(chroma.white[1]) or
        chroma.white[0] <= 0 or chroma.white[1] <= 0 or chroma.white[0] + chroma.white[1] >= 1) return error.Invalid;
    for (definition.gamma) |value| if (!std.math.isFinite(value) or value < 0.01 or value > 100) return error.Invalid;
    const values = [11]f64{ chroma.white[0], chroma.white[1], chroma.red[0], chroma.red[1], chroma.green[0], chroma.green[1], chroma.blue[0], chroma.blue[1], definition.gamma[0], definition.gamma[1], definition.gamma[2] };
    memory.failed = false;
    memory.error_code = 0;
    defer memory.used = 0;
    var size: u32 = @intCast(output.len);
    if (r4gfx_icc_generate(memory, @intFromBool(definition.gray), @intFromBool(definition.srgb_curve), &values, output.ptr, &size) != 0) return if (memory.failed) error.Memory else error.Unsupported;
    return size;
}
pub fn builtin(memory: *Arena, kind: u32, output: []u8) Error!usize {
    if (memory.used != 0 or memory.active != null or output.len > max_profile_bytes or kind > 2) return error.Invalid;
    if (overlaps(@intFromPtr(output.ptr), output.len, @intFromPtr(memory.bytes.ptr), memory.bytes.len) or
        overlaps(@intFromPtr(output.ptr), output.len, @intFromPtr(memory), @sizeOf(Arena)) or
        overlaps(@intFromPtr(memory), @sizeOf(Arena), @intFromPtr(memory.bytes.ptr), memory.bytes.len)) return error.Alias;
    memory.failed = false;
    memory.error_code = 0;
    defer memory.used = 0;
    var size: u32 = @intCast(output.len);
    if (r4gfx_icc_builtin(memory, kind, output.ptr, &size) != 0) return if (memory.failed) error.Memory else error.Unsupported;
    return size;
}
